tarofreighter/wallet.go

```diff
@@ -0,0 +1,644 @@
+	// spend in order to pay the given address.
+	FundAddressSend(ctx context.Context,
+		receiverAddr address.Taro) (*taropsbt.VPacket,
+		*commitment.TaroCommitment, error)
+
+	// FundPacket funds a virtual transaction, selecting assets to spend
+	// in order to pay the given recipient. The selected input is then added
+	// to the given virtual transaction.
+	FundPacket(ctx context.Context, desc *taroscript.RecipientDescriptor,
+		vPkt *taropsbt.VPacket) (*commitment.TaroCommitment, error)
+
+	// SignVirtualPacket signs the virtual transaction of the given packet.
+	SignVirtualPacket(vPkt *taropsbt.VPacket) error
+
+	// AnchorVirtualTransactions creates a BTC level anchor transaction that
+	// anchors all the virtual transactions of the given packets. This
+	// method returns both the funded anchor TX with all the output
+	// information intact for later exclusion proof creation, and the fully
+	// signed and finalized anchor TX along with the total amount of sats
+	// paid in chain fees by the anchor TX
+	AnchorVirtualTransactions(ctx context.Context,
+		feeRate chainfee.SatPerKWeight,
+		inputCommitments []*commitment.TaroCommitment,
+		vPkts []*taropsbt.VPacket) (*AnchorTransaction, error)
+}
+
+// WalletConfig holds the configuration for a new Wallet.
+type WalletConfig struct {
+	// CoinSelector is the interface used to select input coins (assets)
+	// for the transfer.
+	CoinSelector CommitmentSelector
+
+	// KeyRing is used to generate new keys throughout the transfer
+	// process.
+	KeyRing KeyRing
+
+	// Signer implements the Taro level signing we need to sign a virtual
+	// transaction.
+	Signer Signer
+
+	// TxValidator allows us to validate each Taro virtual transaction we
+	// create.
+	TxValidator taroscript.TxValidator
+
+	// Wallet is used to fund+sign PSBTs for the transfer transaction.
+	Wallet WalletAnchor
+
+	// ChainParams is the chain params of the chain we operate on.
+	ChainParams *address.ChainParams
+}
+
+// AssetWallet is an implementation of the Wallet interface that can create
+// virtual transactions, sign them and commit them on-chain.
+type AssetWallet struct {
+	cfg *WalletConfig
+}
+
+// NewAssetWallet creates a new AssetWallet instance from the given
+// configuration.
+func NewAssetWallet(cfg *WalletConfig) *AssetWallet {
+	return &AssetWallet{
+		cfg: cfg,
+	}
+}
+
+// FundAddressSend funds a virtual transaction, selecting assets to spend in
+// order to pay the given address.
+//
+// NOTE: This is part of the Wallet interface.
+func (f *AssetWallet) FundAddressSend(ctx context.Context,
+	receiverAddr address.Taro) (*taropsbt.VPacket,
+	*commitment.TaroCommitment, error) {
+
+	// We start by creating a new virtual transaction that will be used to
+	// hold the asset transfer. Because sending to an address is always a
+	// non-interactive process, we can use this function that always creates
+	// a change output.
+	vPkt := taropsbt.FromAddress(&receiverAddr)
+
+	desc := &taroscript.RecipientDescriptor{
```

Comment:
Is there a reason we don't populate the script key from the address as well?

Reply:
I think because this is described in the `vPSBT` package?

Reply:
Because we only use the descriptor to fund and then verify the inputs to send to the address. So the recipient's script key isn't needed. I removed it from the struct and renamed it to `FundingDescriptor` to make it a bit more clear.